components/eamxx/src/physics/shoc/impl/shoc_assumed_pdf_impl.hpp

```diff
@@ -229,6 +230,27 @@ void Functions<S,D>::shoc_assumed_pdf(
       const auto index_range = ekat::range<IntSmallPack>(k*Spack::n);
       const Smask active_entries = (index_range < nlev);
 
+      // Do NaN Checking here
+      const Smask is_nan_Tl1_1 = (Tl1_1 != Tl1_1) && active_entries;
+      const Smask is_nan_Tl1_2 = (Tl1_2 != Tl1_2) && active_entries;
+      if (is_nan_Tl1_1.any() || is_nan_Tl1_2.any()) {
+        printf("WARNING: NaN Detected in Tl1_1 or Tl1_2! \n");
+	for (int i=0; i<is_nan_Tl1_1.n; i++) {
+          if (is_nan_Tl1_1[i] || is_nan_Tl1_2[i]) {
+	    printf("Tl1 NaN Detected: Tl1_1, Tl1_2: \n");
```

Comment:
I think this set of `printf` statements might allow for multiple ranks to be writing at the same time which may make it confusing to dechiper the warning messages.  Would an `osstream` be the better approach here?  @bartgol do you have any thoughts?

Reply:
It's technically possible. Putting everything together in one printf statement may help. However, ostream cannot be used, since this fcn is to be used on device. If you really want to avoid ranks overlap, you need to use a single printf statement. That doesn't have to make the code ugly, since compile time C strings are cat-ed across lines. So this works:

```
printf ("Tl1 NaN Detected: Tl1_1, Tl1_2: \n"
         "%4.9f,  %4.9f \n"
         "Tl1 NaN Detected, thetal, qw, temp, pressure, thl_sec, qw_sec, w2sec, lev: \n"
         " %4.9f, %2.9f, %4.9f, %8.9f, %4.9f, %2.9f, %4.9f %d \n"
         ...
         " %2.9f, %2.9f, %2.9f, %2.9f \n",
        Tl1_1, Tl1_2,
        thl_first[i], qw_first[i], Tl1_g[i], pval[i], thlsec[i], qwsec[i], w2sec[i], k,
        ...
        qw1_1[i], qw1_2[i], qw2_1[i], qw2_2[i]);
```
Notice *no commas* between consecutive C strings. The compiler splices them together right away.

Reply:
Thanks much, @bartgol, this is very helpful.